### DISCUSSION TITLE
Purging npm component libraries

### DISCUSSION THREAD

**Discussion by samrobbins85** [tailwindlabs/tailwindcss#3267] [Ideas]
At the moment I can't see a particularly good way to publish a component library with tailwind. I have two solutions that have worked for ensuring the classes get included
1. Run tailwind in the component library, to get a CSS file, and include this file in the head of the application
2. In the purging configuration for Tailwind, point it at the folder in `node_modules`

Option 1 has two problems, it introduces duplicate code, and can lead to unexpected behaviour due to the order classes are included.

Option 2 is much better, but feels a bit like a hack

Two nicer approaches would be:
1. Have a key for `component_libraries` (or something like that) in the tailwind config, where you can require the library, and it'll purge it for you. Basically syntactic sugar on top of option 2, but would feel like a proper way to do it.
2. Have a tailwind command that would allow you to get a list of classes to include that you could run on the component library. This could then be imported in some way into the tailwind configuration. This gives the benefit of only running once as the classes needed won't change for an external library.

The issue with both of these is that they will include the classes for unused components, but I can't really think of a way around that without crawling your main code for imports, which seems like a lot of work.

**Comment by maiertech** [comment]
I came across [this gist](https://gist.github.com/somebody32/48277662e55dd450d542724bc989c62d), which shows some code of how you could implement option 2.

**Reply by samrobbins85** [reply]
Thanks! That looks great!

**Comment by maiertech** [comment]
@samrobbins85 Are you aware of any React component libraries that are trying to come up with their own way of implementing option1 or option 2?

**Reply by samrobbins85** [reply]
Not that I'm aware of, they all seem to be just giving you the code to use in your own project like Tailwind UI

**Comment by rtkaaho** [comment]
I just made this post, https://github.com/tailwindlabs/tailwindcss/discussions/3747, with a similar issue. Idea being, create a component library using tailwind classnames and then have the consuming application be responsible for purging. However, like you mentioned, there is no way to tell purgecss to not purge the component library classNames.

**Reply by rtkaaho** [reply]
Found a different solution to this that still requires tailwind to be a dep but is documented here: #3747

**Comment by unrevised6419** [comment]
Would be great to have some webpack plugin. Because webpack knows all the imports, it can deduce all the paths that tailwind needs to purge.

**Reply by IzioDev** [reply]
Or a rollup plugin ;)

**Comment by jquense** [comment]
a related suggestion: https://github.com/tailwindlabs/tailwindcss/discussions/5208
